src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/NamingStyle.cs

```diff
@@ -0,0 +1,289 @@
+                        failureReason = string.Format(FeaturesResources.NamingViolation_WordsCannotContainLowercaseLetters, string.Join(", ", violations));
+                        return false;
+                    }
+                case Capitalization.AllLower:
+                    if (words.SelectMany(w => w.ToCharArray()).All(c => char.IsLower(c)))
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        var violations = words.Where(w => !w.ToCharArray().All(c => char.IsLower(c)));
+                        failureReason = string.Format(FeaturesResources.NamingViolation_WordsCannotContainUppercaseLetters, string.Join(", ", violations));
+                        return false;
+                    }
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private string CreateCompliantNameDirectly(string name)
+        {
+            EnsureNonNullProperties();
+
+            bool addPrefix = !name.StartsWith(Prefix);
+            bool addSuffix = !name.EndsWith(Suffix);
+
+            name = addPrefix ? (Prefix + name) : name;
+            name = addSuffix ? (name + Suffix) : name;
+
+            return FinishFixingName(name);
+        }
+
+        public IEnumerable<string> MakeCompliant(string name)
+        {
+            var name1 = CreateCompliantNameReusingPartialPrefixesAndSuffixes(name);
+            yield return name1;
+
+            var name2 = CreateCompliantNameDirectly(name);
+            if (name2 != name1)
+            {
+                yield return name2;
+            }
+        }
+
+        private string CreateCompliantNameReusingPartialPrefixesAndSuffixes(string name)
+        {
+            EnsureNonNullProperties();
+
+            name = EnsurePrefix(name);
+            name = EnsureSuffix(name);
+
+            return FinishFixingName(name);
+        }
+
+        private string FinishFixingName(string name)
+        {
+            // Edge case: prefix "as", suffix "sa", name "asa"
+            if (Suffix.Length + Prefix.Length >= name.Length)
+            {
+                return name;
+            }
+
+            name = name.Substring(Prefix.Length, name.Length - Suffix.Length - Prefix.Length);
+            IEnumerable<string> words = new[] { name };
+            if (!string.IsNullOrEmpty(WordSeparator))
+            {
+                words = name.Split(new[] { WordSeparator }, StringSplitOptions.RemoveEmptyEntries);
+            }
+
+            words = ApplyCapitalization(words);
+
+            return Prefix + string.Join(WordSeparator, words) + Suffix;
+        }
+
+        private string EnsureSuffix(string name)
+        {
+            // If the name already ends with any prefix of the Suffix, only append the suffix of
+            // the Suffix not contained in the longest such Suffix prefix.
```

Comment:
It's perfectly correct & descriptive, I'm not sure what more you want from me... :wink: